File: 270-ClosestBinarySearchTreeValue/270-ClosestBinarySearchTreeValue.py
# Last updated: 4/29/2026, 10:31:56 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        node = root
        best_value = root.val
        best_diff = abs(root.val - target)
        while node:
            diff = abs(node.val - target)
            if diff < best_diff:
                best_value = node.val
                best_diff = diff
            if node.val == target:
                return node.val
            elif node.val < target:
                node = node.right
            else:
                node = node.left
        return best_value
                
                

File: 270-ClosestBinarySearchTreeValue/270-ClosestBinarySearchTreeValue.js
// Last updated: 4/29/2026, 10:32:01 AM
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {number}
 */
var closestValue = function(root, target) {
    if(!root) {
        return null;
    }
    const obj = {
        minDiff: Math.abs(root.val - target),
        value: root.val,
    }
    dfs(root, target, obj);
    return obj.value;
};

function dfs(node, target, obj) {
    const diff = Math.abs(node.val - target);
    if(diff < obj.minDiff) {
        obj.minDiff = diff;
        obj.value = node.val;
    }
    if(node.val < target) {
        if(node.right) {
            dfs(node.right, target, obj);
        }
    } else if(node.val > target) {
        if(node.left) {
            dfs(node.left, target, obj);
        }
    } else {
        obj.minDiff = 0;
        obj.value = node.val;
    }
}